src/domain/session/room/timeline/deserialize.js

```diff
@@ -0,0 +1,285 @@
+    return new MessageBody(html, parts);
+}
+
+import parse from '../../../../../lib/node-html-parser/index.js';
+
+export class HTMLParseResult {
+    constructor(bodyNode) {
+        this._bodyNode = bodyNode;
+    }
+
+    get rootNodes() {
+        return this._bodyNode.childNodes;
+    }
+
+    getChildNodes(node) {
+        return node.childNodes;
+    }
+
+    getAttributeNames(node) {
+        return node.getAttributeNames();
+    }
+
+    getAttributeValue(node, attr) {
+        return node.getAttribute(attr);
+    }
+
+    isTextNode(node) {
+        return !node.tagName;
+    }
+
+    getNodeText(node) {
+        return node.text;
+    }
+
+    isElementNode(node) {
+        return !!node.tagName;
+    }
+
+    getNodeElementName(node) {
+        return node.tagName;
+    }
+}
+
+const platform = {
+    parseHTML: (html) => new HTMLParseResult(parse(html))
+};
+
+export function tests() {
+    function test(assert, input, output) {
+        assert.deepEqual(parseHTMLBody({ mediaRepository: null, platform }, input), new MessageBody(input, output));
+    }
+
+    return {
+        "Text only": assert => {
+            const input = "This is a sentence";
+            const output = [new TextPart(input)];
+            test(assert, input, output);
+        },
+        "Text with inline code format": assert => {
+            const input = "Here's <em>some</em> <code>code</code>!";
+            const output = [
+                new TextPart("Here's "),
+                new FormatPart("em", [new TextPart("some")]),
+                new TextPart(" "),
+                new FormatPart("code", [new TextPart("code")]),
+                new TextPart("!")
+            ];
+            test(assert, input, output);
+        },
+        "Text with ordered list with no attributes": assert => {
+            const input = "<ol><li>Lorem</li><li>Ipsum</li></ol>";
+            const output = [
+                new ListBlock(1, [
+                    [ new TextPart("Lorem") ],
+                    [ new TextPart("Ipsum") ]
+                ])
+            ];
+            test(assert, input, output);
+        },
+        "Text with ordered list starting at 3": assert => {
+            const input = '<ol start="3"><li>Lorem</li><li>Ipsum</li></ol>';
+            const output = [
+                new ListBlock(3, [
+                    [ new TextPart("Lorem") ],
+                    [ new TextPart("Ipsum") ]
+                ])
+            ];
+            test(assert, input, output);
+        },
+        "Text with unordered list": assert => {
+            const input = '<ul start="3"><li>Lorem</li><li>Ipsum</li></ul>';
+            const output = [
+                new ListBlock(null, [
+                    [ new TextPart("Lorem") ],
+                    [ new TextPart("Ipsum") ]
+                ])
+            ];
+            test(assert, input, output);
+        },
+        /* Doesnt work: HTML library doesn't handle <pre><code> properly.
```

Comment:
the `code` element becomes text. Thus, we end up with `<code>3+x</code>` instead of `3+x`. This make `parseCode` fail, because it expects a `code` element, and not a text node.

Reply:
right, it has special handling to treat everything in `<pre>` as character data... not sure if that's what the spec says, but :man_shrugging: